core/org.wso2.carbon.tomcat.ext/src/main/java/org/wso2/carbon/tomcat/ext/valves/RequestCorrelationIdValve.java

```diff
@@ -290,26 +289,50 @@ private void disAssociateFromThread() {
      *
      * @param queryReceived      Query received in request
      * @param queryConfigured    Query configured in the valve
-     * @param httpServletRequest Request received
+     * @param queryValue         Query value received in request
      * @param correlationIdName  Correlation Id
      * @return A map which contains a query and value that should be associated to the thread
      */
     private Map<String, String> getQueryCorrelationIdValue(String queryReceived, String queryConfigured,
-                                                           HttpServletRequest httpServletRequest,
-                                                           String correlationIdName) {
+                                                           String queryValue, String correlationIdName) {
+
         Map<String, String> queryToAssociate = new HashMap<>();
 
         if (StringUtils.isEmpty(queryReceived) || !StringUtils.equalsIgnoreCase(queryReceived, queryConfigured)) {
             return queryToAssociate;
         }
 
-        String queryValue = httpServletRequest.getParameter(queryReceived);
         if (StringUtils.isNotEmpty(queryValue)) {
             queryToAssociate.put(correlationIdName, queryValue);
         }
         return queryToAssociate;
     }
 
+    /**
+     * Get query parameters from the query string
+     *
+     * @param queryString      Query String
+     * @return A map which contains a name and value of each query param
+     */
+    private Map<String, String[]> getQueryParameters(String queryString) {
+
+        Map<String, String[]> queryParameters = new HashMap<>();
+
+        if (StringUtils.isEmpty(queryString)) {
+            return queryParameters;
+        }
+
+        String[] parameters = queryString.split(QUERY_SEPERATOR);
+
+        for (String parameter : parameters) {
+            String[] keyValuePair = parameter.split(EQUAL);
+            String[] values = queryParameters.get(keyValuePair[0]);
+            values = (String[]) ArrayUtils.add(values, keyValuePair.length == 1 ? "" : keyValuePair[1]);
```

Comment:
What happens to the variable obtained at previous line, it is overwritten.
Ideally it should be appended .
Hence change the String[] to List<String> so that new values can be appended, to cater multiple values like 
url?a=val1&a=val2

Reply:
In such scenarios we will associate both values with "a" separated by comma.